context/app/static/js/components/detailPage/provenance/utils.ts

```diff
@@ -21,11 +42,35 @@ export function nonDestructiveMerge<T extends Record<string, unknown>, U extends
 ): T {
   const merged = { ...a } as Record<string, unknown>;
   Object.entries(b).forEach(([key, value]) => {
+    // Check for duplicate provenance edges by comparing entity and activity IDs
+    if (value && typeof value === 'object' && 'prov:entity' in value && 'prov:activity' in value) {
+      if (
+        Object.values(merged).some(
+          // @ts-expect-error -- We're specifically comparing these keys and have confirmed they exist in the value object
+          (v) => v['prov:entity'] === value['prov:entity'] && v['prov:activity'] === value['prov:activity'],
+        )
+      ) {
+        return;
+      }
+    }
+
     if (merged[key]) {
-      // Confirm that properties are actually different
+      // Heuristic to check if the objects are the same
       if (JSON.stringify(merged[key]) === JSON.stringify(value)) {
         return;
       }
+
+      if (typeof merged[key] === 'object' && typeof value === 'object') {
+        // Perform a more in-depth check to see if objects are the same when excluding certain fields
+        const difference = findDifference(merged[key] as object, value!).filter(
```

Comment:
Forgive my ignorance, but what does the righthand `!` operator do?

Reply:
The `!` operator asserts that the value is non-null; despite having the `typeof value === 'object'` type guard above, this was necessary to avoid it being treated as null in the moment.

I've adjusted the guard statement to be more strict instead.